lib/va_healthcare_messaging/client.rb

```diff
@@ -0,0 +1,184 @@
+require 'va_healthcare_messaging/configuration'
+require 'va_healthcare_messaging/parser'
+require 'va_healthcare_messaging/api/sessions'
+
+module VaHealthcareMessaging
+  #####################################################################################################################
+  ## Client
+  ## Core class responsible for api interface operations with MHV.
+  #####################################################################################################################
+  class Client
+    include VaHealthcareMessaging::API::Sessions
+    include VaHealthcareMessaging::API::TriageTeams
+    include VaHealthcareMessaging::API::Folders
+    include VaHealthcareMessaging::API::Messages
+
+    REQUEST_TYPES = %i(get post delete).freeze
+    USER_AGENT = 'Vets.gov Agent'
+    BASE_REQUEST_HEADERS = {
+      'Accept' => 'application/json',
+      'Content-Type' => 'application/json',
+      'User-Agent' => USER_AGENT
+    }.freeze
+
+    attr_reader :config, :session
+
+    ###################################################################################################################
+    ## initialize
+    ## Intializes the client with both configuration and a session. The configuration contians both the host URL and
+    ## the app_token used to validate the app. The session contains at a minimum the correlation id for the user (if
+    ## no session has yet been established, or has expired).
+    ###################################################################################################################
+    def initialize(config:, session:)
+      @config = config.is_a?(Hash) ? VaHealthcareMessaging::Configuration.new(config) : config
+      @session = session.is_a?(Hash) ? VaHealthcareMessaging::ClientSession.new(session) : session
+
+      raise ArgumentError, 'config is invalid' unless @config.is_a?(Configuration)
+      raise ArgumentError, 'session is invalid' unless @session.valid?
+    end
+
+    ###################################################################################################################
+    ## authenticate
+    ## Gets a valid session token from MHV.
+    ###################################################################################################################
+    def authenticate
+      @session = get_session
+    end
+
+    private
+
+    ###################################################################################################################
+    ## perform
+    ## Communicates with an MHV endpoint.
+    ##
+    ## method:  HTTP method (get, post, ...)
+    ## path:    The MHV endpoint path (session, folders, etc) appended to the MHV URL.
+    ## params:  HTTP variable-value pairs as a hash
+    ## headers: HTTP headers.
+    ##
+    ## Action is
+    ## response = (get, post, ...) -> request -> connection.(get, post, ...) { }.env
+    ###################################################################################################################
```

Comment:
if we need this type of comment, lets try to make it rdoc style similar to what prescriptions_controller does.